### DISCUSSION TITLE
filename=true portability improvements

### DISCUSSION THREAD

**Discussion by davetapley** [duckdb/duckdb#9530] [Ideas]
My use case:
I've got a bunch of `.json` files whose file name contains an ID (e.g. `123.json`), and I'd like to get the ID in a portable way.

---

Per [the docs](https://duckdb.org/docs/data/multiple_files/overview#filename) we can `filename=true` on e.g. `read_json`.

It appears the the glob is portable insomuch as I can use either `/` or `\` as a separator on Windows or Linux, and both work
(i.e. it's automatically converting to the correct one under the hood), nice ✅ 

Unfortunately, the `filename` I get back is always OS specific (i.e. `/` on Linux and `\` on Windows).
Not _terrible_ since I can `string_split_regex(filename[0:-6], '[/\\]')[-1]`, but:

**Idea 1:** could the returned path's separator always be consistent with the one which was given in the glob?

---

Of course, that's a bit of an XY problem and what I really want is:

**Idea 2:** can we get some path handling functions?

A new path type might be overkill, but if we could make e.g.:
1. `path_parse(VARCHAR)` which is just `string_split_regex(VARCHAR, '[/\\]')`
2. `path_name(VARCHAR[])` which is `[-1]`
3. `path_stem(VARCHAR[])` which is `string_split(VARCHAR[][-1], '.')[0]`

Thoughts?

**Comment by samansmink** [comment]
Having some clean way to handle paths in a platform (in)dependent way seems reasonable.
Im not sure if returning the filename always as a forward slash is the way to go, but having some scalar functions that can split paths automatically seems nice, this might also be useful in cases where paths are found in datasets

My suggestion would be that we add some platform dependent scalar functions here:
```
# Returns a list of the parsed path
VARCHAR[] parse_path(VARCHAR path [, VARCHAR separator])

# Returns the filename (trim_ext defaults to false)
VARCHAR parse_filename(VARCHAR path [, BOOLEAN trim_extension] [, VARCHAR separator])

# Returns the top level directory name
VARCHAR parse_dirname(VARCHAR path [, VARCHAR separator])
```

then the optional separator param would be either: `system`, `both_slash`, `forward_slash`, `backward_slash`
the default can then be `both_slash`

**Comment by chrisiou** [comment]
These functions have been introduced in PR [#10208](https://github.com/duckdb/duckdb/pull/10208). Additionally, the `parse_dirpath()` has been added, which:

```
# Returns the head of the path (the pathname until the last slash) similarly to Python's os.path.dirname
VARCHAR parse_dirpath(VARCHAR path [, VARCHAR separator])
```

They will be available in the upcoming release. You can refer to the [Release Dates](https://duckdb.org/dev/release-dates) for more information.
